File: typescript/packages/core/src/commands/builtin/generic_bind/builders/gunzip.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { gunzipGeneric } from '../../generic/gunzip.ts'
import { type Builder, requireOp, resolveGlobOf } from '../adapter.ts'

export const GUNZIP_BUILDER: Builder = {
  name: 'gunzip',
  write: true,
  fn: async (ops, accessor, paths, _texts, opts) => {
    const idx = opts.index ?? undefined
    const write = requireOp(ops.write, 'write')
    const unlink = requireOp(ops.unlink, 'unlink')
    const resolved = paths.length > 0 ? await resolveGlobOf(ops)(accessor, paths, idx) : []
    return gunzipGeneric(
      resolved,
      opts,
      (p) => ops.readStream(accessor, p, idx),
      (p, d) => write(accessor, p, d),
      (p) => unlink(accessor, p),
    )
  },
}

File: python/mirage/commands/builtin/generic_bind/builders/gunzip.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from functools import partial

from mirage.accessor.base import Accessor
from mirage.commands.builtin.generic.gunzip import gunzip_generic
from mirage.commands.builtin.generic_bind.adapter import (Builder, CommandIO,
                                                          Operation, bound_op)
from mirage.commands.config import CommandOpts
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


async def gunzip(ops: CommandIO, accessor: Accessor, paths: list[PathSpec],
                 texts: list[str],
                 opts: CommandOpts) -> tuple[ByteSource | None, IOResult]:
    resolved = await ops.resolve_glob(accessor, paths,
                                      opts.index) if paths else []
    return await gunzip_generic(
        resolved, list(texts), opts,
        bound_op(ops.read_bytes, accessor, opts.index),
        partial(ops.require(Operation.WRITE), accessor),
        partial(ops.require(Operation.UNLINK), accessor))


BUILDER = Builder('gunzip', gunzip, write=True)
